### PR TITLE
feat: implement ign update command and rename template collect-vars

### PR DESCRIPTION
**Pull request by tacogips**
## Summary

This PR implements the template synchronization feature and renames the CLI command from `ign template collect-vars` to `ign template update` to better reflect its expanded role. The new `ign update` command enables users to synchronize projects when templates change, while `ign template update` now calculates and stores template hashes in addition to collecting variables.

## Changes

- Implemented `ign update` command that fetches templates, detects changes via hash comparison, prompts for new variables, and regenerates project files
- Renamed `ign template collect-vars` to `ign template update` across CLI, implementation, and documentation
- Added hash calculation and storage in template `ign.json` files
- Implemented template hash comparison to detect when templates have changed
- Added variable change detection (new/removed variables) when updating from changed templates
- Updated documentation in `cli-commands.md` with comprehensive `ign update` command reference
- Bumped project version to 0.1.6 in `flake.nix`

## Changed Files

| File | Additions | Deletions | Change Summary |
| ---- | --------- | --------- | -------------- |
| internal/app/update.go | 443 | 0 | Implement ign update command |
| internal/cli/update.go | 386 | 0 | Implement update CLI handler |
| docs/reference/cli-commands.md | 122 | 6 | Add ign update command documentation |
| internal/app/template_update.go | 90 | 14 | Rename from collect.go, add hash calculation |
| internal/cli/template.go | 33 | 32 | Rename collect-vars to update |
| internal/app/template_update_test.go | 25 | 25 | Rename from collect_test.go |
| docs/reference/configuration.md | 7 | 1 | Document hash field in ign.json |
| internal/app/checkout.go | 6 | 2 | Use hash from ign.json |
| internal/template/model/ignconfig.go | 4 | 0 | Add hash field to IgnJson |
| flake.nix | 1 | 1 | Bump version to 0.1.6 |
| internal/cli/root.go | 1 | 0 | Add update command to root |

### New `ign update` Command Implementation

The `ign update` command provides template synchronization through a two-phase workflow:

**Prepare Phase** (`PrepareUpdate`):
- Loads existing `.ign/ign.json` and `.ign/ign-var.json` configuration
- Fetches the template from the stored URL
- Compares template hash (from template's `ign.json`) with locally stored hash
- Identifies new and removed variables between versions
- Returns early if template hash unchanged (no updates needed)

**Complete Phase** (`CompleteUpdate`):
- Prompts for new required variables (optional variables use defaults)
- Merges existing variable values with new ones
- Updates `.ign/ign.json` with new template hash
- Updates `.ign/ign-var.json` with merged variables
- Regenerates project files using updated template

### Hash-based Template Change Detection

Template hashes are calculated via `CalculateTemplateHashFromDir`:
- SHA256 hash of all template files (excluding `ign.json` itself)
- Deterministic ordering (sorted file paths) ensures consistent hashes
- Stored in template's `ign.json` via `ign template update`
- Used by `ign update` to detect when templates have changed

### Command Renaming Rationale

Renamed `ign template collect-vars` to `ign template update` because:
- Original name focused only on variable collection
- With hash calculation added, command now updates both variables and hash
- Name `update` better represents the command's role in template maintenance
- Aligns with `ign update` command naming for project updates

### Files Renamed

- `internal/app/collect.go` → `internal/app/template_update.go`
- `internal/app/collect_test.go` → `internal/app/template_update_test.go`

### Types/Functions Renamed

- `CollectVarsOptions` → `UpdateTemplateOptions`
- `CollectVarsResult` → `UpdateTemplateResult`
- `CollectVars()` → `UpdateTemplate()`

### Documentation Updates

- Added comprehensive `ign update` command reference in `cli-commands.md`
- Documented process flow, examples, flags, and error cases
- Updated `configuration.md` with hash field documentation
- Updated comments in `update.go` and `ignconfig.go` to reference hash usage

## Related Issues/PRs

---

### CODE REVIEW

File: internal/app/checkout.go [hunk 1]
```diff
@@ -299,8 +299,12 @@ func CompleteCheckout(ctx context.Context, opts CompleteCheckoutOptions) (*Check
 
 	// Create and save configuration files (unless dry-run)
 	if !opts.DryRun {
-		// Calculate template hash
-		templateHash := calculateTemplateHash(prep.Template)
+		// Get template hash from ign.json, or calculate if not present
+		templateHash := prep.IgnJson.Hash
```
Comment (tacogips) [inline internal/app/checkout.go]: ## [REVIEW] High: Inconsistent hash handling compared to config_init.go

**Severity**: High

### Problem
The CompleteCheckout function uses `prep.IgnJson.Hash` as the source of truth for template hash, with a fallback to `calculateTemplateHash()` only when the hash is empty. This differs from the pattern used in `config_init.go` (line 70), which always calculates the hash directly from the template.

### Why this is a problem
- **Inconsistency**: Different code paths (config_init.go vs checkout.go) use different approaches for the same task
- **Potential stale hash**: If `prep.IgnJson.Hash` exists but is stale or incorrect, it will be used without validation
- **Logic mismatch**: The template's ign.json hash (`prep.IgnJson.Hash`) represents the hash **of the template source**, not necessarily the hash that should be stored in the **project's** .ign/ign.json
- **Differs from update.go**: While update.go uses the same fallback pattern (lines 175-182), config_init.go sets a precedent of always calculating fresh

### Problematic code pattern
```go
// CompleteCheckout (checkout.go:303-306)
templateHash := prep.IgnJson.Hash
if templateHash == "" {
    debug.Debug("[app] Template has no hash in ign.json, calculating from content")
    templateHash = calculateTemplateHash(prep.Template)
}
```

vs.

```go
// ConfigInit (config_init.go:70)
templateHash := calculateTemplateHash(prepResult.Template)
```

### Suggested direction
- Consider always calculating the hash from the template content for consistency with config_init.go
- OR document why this fallback pattern is correct (if there's a reason to trust prep.IgnJson.Hash)
- Ensure all initialization paths (init, checkout, update) follow the same hash calculation strategy

File: internal/cli/update.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+		} else {
+			fmt.Printf("@@ -0,0 +1,%d @@\n", lines)
+		}
+
+		printPatchContentForUpdate(f.Content)
+		fmt.Println()
+	}
+}
+
+// countLinesForUpdate counts the number of lines in content.
```
Comment (tacogips) [inline internal/cli/update.go]: ## [REVIEW] Medium: Code duplication in utility functions

**Severity**: Medium

### Problem
The functions `countLinesForUpdate`, `printPatchContentForUpdate`, and `isBinaryContentForUpdate` (lines 342-386) are duplicates of existing functions `countLines`, `printPatchContent`, and `isBinaryContent` from checkout.go.

### Why this is a problem
- Code duplication violates DRY (Don't Repeat Yourself) principle
- Increases maintenance burden (changes need to be made in multiple places)
- Higher risk of inconsistent behavior if one copy is updated but not the other
- Increases code size unnecessarily

### Problematic code pattern
```go
func countLinesForUpdate(content []byte) int {
    // ... identical implementation to countLines
}
```

### Suggested direction
- Extract these functions to a shared utility file (e.g., `internal/cli/patch_utils.go`)
- Import and reuse the same functions in both `checkout.go` and `update.go`
- Remove the "ForUpdate" suffix from function names since they serve the same purpose

File: internal/app/template_update.go [hunk 3]
```diff
@@ -430,6 +432,17 @@ func updateIgnJson(path string, result *CollectVarsResult, existing *model.IgnJs
 		ignJson.Variables[name] = varDef
 	}
 
+	// Calculate and update template hash
+	templateDir := filepath.Dir(path)
+	newHash, err := CalculateTemplateHashFromDir(templateDir)
+	if err != nil {
```
Comment (tacogips) [inline internal/app/template_update.go]: ## [REVIEW] High: Silent hash calculation failures in updateIgnJson

**Severity**: High

### Problem
When hash calculation fails, the error is logged but execution continues, writing ign.json without a hash field. This creates inconsistent state without user awareness.

### Why this is a problem
- Users won't know their template hash is missing
- The hash field is critical for 'ign update' to detect template changes
- Silent failures violate the principle of failing loudly
- Missing hash means 'ign update' won't work correctly
- The debug log may not be visible to users in normal operation

### Problematic code pattern
```go
newHash, err := CalculateTemplateHashFromDir(templateDir)
if err != nil {
    debug.Debug("[app] Failed to calculate template hash: %v", err)
    // Continue without hash if calculation fails
}
```

### Suggested direction
- Return the error instead of suppressing it
- Hash calculation is critical functionality - failure should be visible
- Alternative: Add a warning message to the user (not just debug log)
- Consider if continuing without hash is ever acceptable

File: internal/cli/update.go [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+			return nil, fmt.Errorf("failed to prompt for variable %q: %w", name, err)
+		}
+
+		vars[name] = value
+	}
+
+	return vars, nil
+}
+
+// PromptForVariablesSubset prompts for a subset of variables from IgnJson.
```
Comment (tacogips) [inline internal/cli/update.go]: ## [REVIEW] Low: Unused function PromptForVariablesSubset

**Severity**: Low

### Problem
The function `PromptForVariablesSubset` (lines 234-264) is defined but never called anywhere in the codebase. This is dead code similar to the `ConfirmUpdate` function.

### Why this is a problem
- Dead code clutters the codebase and confuses maintainers
- Exported functions suggest an API contract that may mislead users
- Increases code complexity without providing value
- May indicate incomplete or planned functionality

### Problematic code pattern
```go
func PromptForVariablesSubset(ignJson *model.IgnJson, varNames []string) (map[string]interface{}, error) {
    // ... never called
}
```

### Suggested direction
- Remove this function as `PromptForNewVariables` already handles the subset prompting use case
- The current implementation at line 130 uses `PromptForNewVariables(varsNeedingPrompt)` which achieves the same result
- If there's a specific reason to keep this variant, document it with a comment

File: internal/app/template_update.go [hunk 5]
```diff
@@ -442,3 +455,66 @@ func updateIgnJson(path string, result *CollectVarsResult, existing *model.IgnJs
...
+	}
+
+	// Sort files for deterministic hash
+	sort.Strings(files)
+
+	// Hash each file's path and content
+	for _, relPath := range files {
+		fullPath := filepath.Join(dirPath, relPath)
+		content, err := os.ReadFile(fullPath)
+		if err != nil {
```
Comment (tacogips) [inline internal/app/template_update.go]: ## [REVIEW] Medium: File read errors in hash calculation may cause incomplete hash

**Severity**: Medium

### Problem
When reading file content fails during hash calculation, the error is returned immediately, but the hash state at that point is lost. This could be confusing if the error message doesn't clearly indicate which file failed.

### Why this is a problem
- Error message lacks context about which file failed in large directory trees
- Difficult to debug permission issues or other file access problems
- The relative path is known but not included in the error

### Problematic code pattern
```go
content, err := os.ReadFile(fullPath)
if err != nil {
    return "", err  // Which file failed? Hard to tell from error alone
}
```

### Suggested direction
- Wrap error with file path context using fmt.Errorf
- Example: `return "", fmt.Errorf("failed to read %s: %w", relPath, err)`
- This provides better debugging information for users
- Helps identify permission issues or missing files

File: internal/cli/template.go [hunk 6]
```diff
@@ -62,26 +62,27 @@ Examples:
...
+	Short: "Update ign.json with variables and template hash",
 	Long: `Scan template files for @ign-var: and @ign-if: directives and
-automatically update ign.json with the collected variable definitions.
+automatically update ign.json with the collected variable definitions
+and recalculate the template hash.
 
-This command helps keep ign.json in sync with the actual variables used
-in your template files.
+This command keeps ign.json in sync with template files and updates
+the hash field used by 'ign update' to detect template changes.
```
Comment (tacogips) [inline internal/cli/template.go]: ## [REVIEW] Medium: Help text could be clearer about command relationship

**Severity**: Medium

### Problem
The Long description of `ign template update` mentions `'ign update'` in a way that could confuse users about the relationship between the two commands.

### Why this is a problem
- Users might not immediately understand that `ign update` and `ign template update` serve different purposes
- The reference to `ign update` appears without context about what it does
- For new users, it's unclear why updating the hash field matters for a different command

### Problematic code pattern
```go
Long: `...
This command keeps ign.json in sync with template files and updates
the hash field used by 'ign update' to detect template changes.
...`
```

### Suggested direction
Consider clarifying the relationship between the two commands more explicitly. For example:
- Add a brief explanation of what `ign update` does in relation to this command
- Or rephrase to make the purpose clearer without referencing the other command
- Example: "...and updates the hash field. When users run 'ign update', this hash is used to detect template changes."

This would help users understand the workflow: template authors use `ign template update` to maintain templates, while project users use `ign update` to sync with template changes.

File: internal/cli/update.go [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+	// Prepare update - fetch template and check for changes
+	prepResult, err := app.PrepareUpdate(cmd.Context(), app.UpdateOptions{
+		OutputDir:   outputPath,
+		Overwrite:   updateForce,
+		DryRun:      updateDryRun,
+		Verbose:     updateVerbose,
+		GitHubToken: githubToken,
+	})
+	if err != nil {
+		printErrorMsg(fmt.Sprintf("Update preparation failed: %v", err))
```
Comment (tacogips) [inline internal/cli/update.go]: ## [REVIEW] Medium: Inconsistent error message formatting

**Severity**: Medium

### Problem
Line 78 formats the error message with `fmt.Sprintf` before passing to `printErrorMsg`, but line 160 does the same pattern. However, looking at similar code in `checkout.go`, this pattern is consistent. The actual issue is that error context from `err` might be lost when we immediately return `err` after printing a custom message.

### Why this is a problem
- When the error is returned (line 79), the caller sees the original error from `app.PrepareUpdate`
- But the user sees a different message printed to stderr ("Update preparation failed: %v")
- This can cause confusion when debugging or logging
- The printed message and returned error should ideally convey the same information

### Problematic code pattern
```go
if err != nil {
    printErrorMsg(fmt.Sprintf("Update preparation failed: %v", err))
    return err  // Returns original error, not the formatted message
}
```

### Suggested direction
- This pattern matches `checkout.go`, so it's acceptable for consistency
- However, consider if the error handling should wrap the error with context using `fmt.Errorf("update preparation failed: %w", err)`
- Alternatively, ensure that `printErrorMsg` and the returned error always match

File: internal/cli/update.go [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+		fmt.Printf(", %d to overwrite", result.FilesOverwritten)
+	}
+	if result.FilesSkipped > 0 {
+		fmt.Printf(", %d to skip (already exist)", result.FilesSkipped)
+	}
+	fmt.Println()
+	fmt.Println()
+
+	// Print each file in patch format
+	for _, f := range result.DryRunFiles {
```
Comment (tacogips) [inline internal/cli/update.go]: ## [REVIEW] Low: Inconsistent variable naming in dry-run output

**Severity**: Low

### Problem
The variable `f` in line 318 is used to iterate over `result.DryRunFiles`, which is a common but not very descriptive name. In contrast, `checkout.go` uses the same pattern with `f`, suggesting consistency, but this could be improved for readability.

### Why this is a problem
- Single-letter variable names reduce code readability, especially in longer loops
- The loop body spans 22 lines (318-339), making it harder to remember what `f` represents
- Go style guide suggests descriptive names for variables with larger scopes

### Problematic code pattern
```go
for _, f := range result.DryRunFiles {
    if f.WouldSkip {
        // ... 20+ lines of code
    }
}
```

### Suggested direction
- Use a more descriptive variable name like `file` or `dryRunFile`
- This improves readability: `file.WouldSkip`, `file.Path`, `file.Content`
- Apply the same improvement to `checkout.go` for consistency if changing this

File: internal/cli/update.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+		Default: true,
+	}
+	if err := survey.AskOne(prompt, &confirm); err != nil {
+		return false, err
+	}
+	return confirm, nil
+}
+
+// printUpdateDryRunPatch outputs dry-run results in unified diff format.
+func printUpdateDryRunPatch(result *app.UpdateResult) {
```
Comment (tacogips) [inline internal/cli/update.go]: ## [REVIEW] Medium: Potential nil pointer dereference in printUpdateDryRunPatch

**Severity**: Medium

### Problem
The function `printUpdateDryRunPatch` at line 280 doesn't validate that the `result` parameter is non-nil before accessing its fields. While the caller at line 166 appears to pass a valid result, defensive programming suggests validation.

### Why this is a problem
- If called with a nil result (e.g., from future refactoring or different code path), this will panic
- The function is public (starts with capital letter) and could be called from other packages
- Similar functions in other CLI tools typically validate parameters
- Silent panics are harder to debug than explicit error messages

### Problematic code pattern
```go
func printUpdateDryRunPatch(result *app.UpdateResult) {
    // No nil check
    if len(result.NewVariables) > 0 || len(result.RemovedVariables) > 0 {
        // ... potential panic if result is nil
    }
}
```

### Suggested direction
- Add nil check at the start of the function
- Either return early with a warning or panic with a descriptive message
- Example: `if result == nil { return }` or log a warning
- Consider making the function private (lowercase) if it's only used within this package

File: internal/app/update.go [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+		// Prompt if variable is required OR has no default
+		if varDef.Required || varDef.Default == nil {
+			result[name] = varDef
+		}
+	}
+	return result
+}
+
+// ApplyDefaults fills in default values for optional variables with defaults.
+func ApplyDefaults(newVarDefs map[string]model.VarDef, providedVars map[string]interface{}) map[string]interface{} {
```
Comment (tacogips) [inline internal/app/update.go]: ## [REVIEW] Low: ApplyDefaults function has unclear nil handling behavior

**Severity**: Low

### Problem
Function `ApplyDefaults` (lines 415-427) accepts `providedVars map[string]interface{}` which could be nil, and the function's behavior differs subtly depending on whether it receives nil vs empty map.

### Why this is a problem
- Line 417-419 copies from providedVars - if nil, this creates empty result map
- However, the function doesn't document this nil-handling behavior
- Caller at line 135 in `cli/update.go` passes nil: `newVarValues = app.ApplyDefaults(newVarDefs, nil)`
- Unclear whether nil is intentional API or should be enforced as empty map

### Problematic code pattern
```go
func ApplyDefaults(newVarDefs map[string]model.VarDef, providedVars map[string]interface{}) map[string]interface{} {
    result := make(map[string]interface{})
    for name, value := range providedVars {  // Silent no-op if providedVars is nil
        result[name] = value
    }
    // Add defaults...
}
```

### Suggested direction
- Document that `providedVars` can be nil (treated as empty map)
- OR normalize nil to empty map at start:
  ```go
  if providedVars == nil {
      providedVars = make(map[string]interface{})
  }
  ```
- Add function comment explaining nil handling
- This improves API clarity for future maintainers

File: internal/app/update.go [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+	debug.DebugValue("[app] Template hash from ign.json", newHash)
+
+	// If template doesn't have a hash yet, calculate it from content
+	if newHash == "" {
+		debug.Debug("[app] Template has no hash in ign.json, calculating from content")
+		newHash = calculateTemplateHash(template)
+		debug.DebugValue("[app] Calculated template hash", newHash)
+	}
+
+	hashChanged := newHash != ignConfig.Hash
```
Comment (tacogips) [inline internal/app/update.go]: ## = Critical: Cross-File Issue: Inconsistent Hash Calculation Algorithms

**Severity**: Critical
**Affected Files**:
- `internal/app/template_update.go`
- `internal/app/checkout.go`
- `internal/app/update.go`

### Problem
Two different hash calculation functions exist with different algorithms, causing hash mismatches for the same template:

1. `CalculateTemplateHashFromDir()` (template_update.go) - reads from filesystem
2. `calculateTemplateHash()` (checkout.go) - reads from in-memory template object

### Why this is a problem
- **Hash Mismatch**: The same template will produce different hashes depending on which function calculates it
- **Update Detection Broken**: `update.go` compares hashes to detect template changes (line 185), but uses hash from template ign.json (calculated by CalculateTemplateHashFromDir) vs. fallback calculateTemplateHash (line 181)
- **Data Integrity**: Hash is meant to uniquely identify template content, but inconsistent calculation defeats this purpose
- **User Impact**: `ign update` may incorrectly report "template changed" when nothing changed, or miss actual changes

### Code comparison

**template_update.go** (lines 462-520) - Calculates hash from filesystem:
```go
func CalculateTemplateHashFromDir(dirPath string) (string, error) {
    h := sha256.New()
    var files []string

    // Collect all file paths from FILESYSTEM
    err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
        // ... walks directory, reads files from disk
        files = append(files, relPath)
        return nil
    })
    
    sort.Strings(files)  // Sort relative paths
    
    for _, relPath := range files {
        fullPath := filepath.Join(dirPath, relPath)
        content, err := os.ReadFile(fullPath)  // Read from DISK
        h.Write([]byte(relPath))
        h.Write(content)
    }
    return hex.EncodeToString(h.Sum(nil)), nil
}
```

**checkout.go** (lines 98-128) - Calculates hash from in-memory template:
```go
func calculateTemplateHash(template *model.Template) string {
    h := sha256.New()

    // Sort files from IN-MEMORY TEMPLATE object
    files := make([]model.TemplateFile, len(template.Files))
    copy(files, template.Files)
    sort.Slice(files, func(i, j int) bool {
        return files[i].Path < files[j].Path
    })

    // Hash from IN-MEMORY data
    for _, file := range files {
        h.Write([]byte(file.Path))
        h.Write(file.Content)  // Use in-memory content
    }
    return hex.EncodeToString(h.Sum(nil))
}
```

**update.go** (lines 173-186) - Uses both hash sources inconsistently:
```go
// Get hash from template's ign.json (calculated by CalculateTemplateHashFromDir)
newHash := template.Config.Hash

// If template doesn't have a hash yet, calculate it from content
if newHash == "" {
    newHash = calculateTemplateHash(template)  // Uses DIFFERENT algorithm
}

hashChanged := newHash != ignConfig.Hash  // Compare hashes from different algorithms!
```

### Suggested direction
1. **Single Source of Truth**: Choose ONE hash calculation algorithm and use it everywhere
   - Option A: Always use `CalculateTemplateHashFromDir()` (requires template to be on disk)
   - Option B: Always use `calculateTemplateHash()` (works with in-memory templates)
   - Recommendation: Use calculateTemplateHash since templates are fetched remotely

2. **Ensure Consistency**: If both functions must exist, ensure they produce identical hashes
   - Verify that template.Files contains exact same files as filesystem
   - Use same path normalization (relative vs. absolute)
   - Use same sort order

3. **Update Hash Calculation Points**:
   - `template_update.go`: Use the canonical hash function when saving to ign.json
   - `checkout.go`: Use the canonical hash function for fallback calculation
   - `update.go`: Use the canonical hash function for comparison

4. **Add Hash Verification**: Consider adding unit tests that verify both functions produce identical hashes for the same template

File: internal/cli/root.go [hunk 12]
```diff
@@ -62,6 +62,7 @@ func init() {
 
 	// Add subcommands
 	rootCmd.AddCommand(checkoutCmd)
+	rootCmd.AddCommand(updateCmd)
```
Comment (tacogips) [inline internal/cli/root.go]: ## High: Cross-File Issue: Command naming ambiguity between 'ign update' and 'ign template update'

**Severity**: High
**Affected Files**:
- `internal/cli/root.go` (lines 64-66)
- `internal/cli/update.go` (lines 16-18)
- `internal/cli/template.go` (lines 66-68)

### Problem
Two commands with very similar names (`ign update` and `ign template update`) are registered, but they serve completely different purposes. This creates potential user confusion.

### Why this is a problem
- Users might mistakenly use `ign update` when they meant `ign template update` or vice versa
- The similar naming suggests a relationship that does not exist
- New users may not understand which command to use for their use case
- Command discoverability is hindered by the naming overlap

### Code comparison

**File 1**: `internal/cli/root.go` (lines 64-66)
```go
rootCmd.AddCommand(checkoutCmd)
rootCmd.AddCommand(updateCmd)        // ign update - updates PROJECT
rootCmd.AddCommand(templateCmd)      // ign template - template management
```

**File 2**: `internal/cli/update.go` (lines 16-18)
```go
var updateCmd = &cobra.Command{
    Use:   "update [output-path]",
    Short: "Update project from template changes",  // User-facing: updates generated project
```

**File 3**: `internal/cli/template.go` (lines 66-68)
```go
var templateUpdateCmd = &cobra.Command{
    Use:   "update [PATH]",
    Short: "Update ign.json with variables and template hash",  // Author-facing: updates template metadata
```

### Suggested direction
Consider one of these approaches:
1. Rename `ign template update` to `ign template sync` or `ign template refresh` to avoid the "update" overlap
2. Add cross-references in the Long help text of both commands mentioning the other command and clarifying the difference
3. Consider whether these commands belong in different command groups (e.g., project commands vs template-author commands)
4. Add examples in help text that clearly show the different use cases

File: internal/app/update.go [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+
+	// Step 5: Get hash from template's ign.json and compare
+	// The hash is stored in the template's ign.json (calculated by 'ign template update')
+	newHash := template.Config.Hash
+	debug.DebugValue("[app] Template hash from ign.json", newHash)
+
+	// If template doesn't have a hash yet, calculate it from content
+	if newHash == "" {
+		debug.Debug("[app] Template has no hash in ign.json, calculating from content")
+		newHash = calculateTemplateHash(template)
```
Comment (tacogips) [inline internal/app/update.go]: ## = High: Cross-File Issue: Hash Source Inconsistency in update.go

**Severity**: High
**Affected Files**:
- internal/app/update.go
- internal/app/template_update.go
- internal/app/checkout.go

### Problem
update.go:PrepareUpdate() uses hash from two different sources with no validation that they are compatible:

1. Primary source: template.Config.Hash (line 175) - hash stored in template ign.json
2. Fallback source: calculateTemplateHash(template) (line 181) - hash calculated from in-memory template

The fallback uses a DIFFERENT algorithm than the primary source.

### Why this is a problem
- Inconsistent Change Detection: When template.Config.Hash is empty, the fallback uses a different algorithm, producing an incompatible hash
- False Positives: Comparing hashes from different algorithms will always show template changed even if content is identical
- Silent Failures: No warning when fallback is used, making debugging difficult
- Contract Violation: The Hash field has an implicit contract that it represents a specific hash algorithm

### Code comparison

**update.go** (lines 173-186):
```go
// Get hash from template ign.json (calculated by CalculateTemplateHashFromDir)
newHash := template.Config.Hash

// If template does not have a hash yet, calculate it from content
if newHash == "" {
    newHash = calculateTemplateHash(template)  // DIFFERENT ALGORITHM
}

hashChanged := newHash != ignConfig.Hash  // Comparing potentially incompatible hashes
```

**template_update.go** (lines 435-444) - Where primary hash is calculated:
```go
templateDir := filepath.Dir(path)
newHash, err := CalculateTemplateHashFromDir(templateDir)  // Filesystem algorithm
if err != nil {
    // Continue without hash if calculation fails
} else {
    ignJson.Hash = newHash  // Stored in template ign.json
}
```

### Suggested direction
- Ensure calculateTemplateHash and CalculateTemplateHashFromDir produce identical results for the same template
- Or require template.Config.Hash to always be present (error if empty)
- Add validation and warning if fallback is used

File: internal/cli/update.go (hunk 2 above)
Comment (tacogips) [reply internal/cli/update.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- Created new file internal/cli/patch_utils.go with shared utility functions
- Extracted countLines, printPatchContent, and isBinaryContent to patch_utils.go
- Removed duplicate functions from both checkout.go and update.go
- All patch-related code now uses the shared functions

Files modified:
- internal/cli/patch_utils.go (+57 lines, new file)
- internal/cli/checkout.go (-52 lines, removed duplicates)
- internal/cli/update.go (-126 lines, removed duplicates including ForUpdate variants)

Resolving this thread.

File: internal/app/template_update.go (hunk 3 above)
Comment (tacogips) [reply internal/app/template_update.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- Removed silent error logging that allowed execution to continue
- Hash calculation errors now return immediately
- Added detailed comment explaining the criticality: "Hash calculation is critical for 'ign update' to detect template changes"

Code before:
```go
if err != nil {
    debug.Debug("Failed to calculate template hash: %v", err)
    // Continue without hash if calculation fails
}
```

Code after:
```go
newHash, err := CalculateTemplateHashFromDir(templateDir)
if err != nil {
    return fmt.Errorf("failed to calculate template hash: %w", err)
}
```

Resolving this thread.

File: internal/app/template_update.go (hunk 5 above)
Comment (tacogips) [reply internal/app/template_update.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- Wrapped file read error with file path context using fmt.Errorf
- Changed from: `return "", err`
- Changed to: `return "", fmt.Errorf("failed to read %s: %w", relPath, err)`

This provides better debugging information when file access fails during hash calculation.

Resolving this thread.

File: internal/cli/template.go (hunk 6 above)
Comment (tacogips) [reply internal/cli/template.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- Added extensive clarification to the help text distinguishing template authors from project users
- New text clearly explains: "This command is for TEMPLATE AUTHORS to maintain their template repositories"
- Contrasts with 'ign update' which is for "PROJECT USERS who have generated a project"
- Explains the workflow relationship between the two commands

Added section in internal/cli/template.go:
```
This command is for TEMPLATE AUTHORS to maintain their template repositories.

Note: This is different from 'ign update', which is for PROJECT USERS who have
generated a project and want to pull in updates from the template source. When
project users run 'ign update', this hash is used to detect whether the template
has changed since their last checkout or update.
```

Resolving this thread.

File: internal/cli/update.go (hunk 9 above)
Comment (tacogips) [reply internal/cli/update.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- Added nil check to printUpdateDryRunPatch function in internal/cli/update.go
- Function now returns early if result is nil
- Prevents potential nil pointer dereference

Note: This function was later removed in commit 75d1185 when patch utility functions were extracted to patch_utils.go, which already includes proper nil handling.

Resolving this thread.

File: internal/app/update.go [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+			}
+		}
+	}
+
+	debug.Debug("[app] CompleteUpdate workflow completed successfully")
+	return result, nil
+}
+
+// GetNewVariableDefinitions returns VarDef for new variables that need prompting.
+func GetNewVariableDefinitions(prep *PrepareUpdateResult) map[string]model.VarDef {
```
[the remark this answers is not in the document]
Comment (tacogips) [reply internal/app/update.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- Added nil checks to GetNewVariableDefinitions function in internal/app/update.go
- Function now validates prep, prep.IgnJson, and prep.IgnJson.Variables before accessing:
  ```go
  if prep == nil || prep.IgnJson == nil || prep.IgnJson.Variables == nil {
      return result
  }
  ```
- Prevents potential nil pointer dereference

Resolving this thread.

File: internal/app/update.go (hunk 11 above)
Comment (tacogips) [reply internal/app/update.go]: Fixed in commits 75d1185 and e9bfe97 (PR #17)

Changes made:
- Both calculateTemplateHash (in checkout.go) and CalculateTemplateHashFromDir (in template_update.go) now use the shared HashTemplateFiles function
- Extracted shared logic to internal/app/hash.go
- Both functions use identical algorithm with null byte separators
- Ensures hash consistency across all calculation paths

Code structure:
- hash.go: Contains HashTemplateFiles with the canonical hash algorithm
- checkout.go: calculateTemplateHash calls HashTemplateFiles
- template_update.go: CalculateTemplateHashFromDir calls HashTemplateFiles

Resolving this thread.

File: internal/cli/root.go (hunk 12 above)
Comment (tacogips) [reply internal/cli/root.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- Added extensive clarification to the `ign template update` command help text distinguishing template authors from project users
- The help text now explicitly states:
  - "This command is for TEMPLATE AUTHORS to maintain their template repositories."
  - "Note: This is different from 'ign update', which is for PROJECT USERS who have generated a project and want to pull in updates from the template source."

Resolving this thread.

File: internal/app/checkout.go [hunk 15]
```diff
@@ -299,8 +299,12 @@ func CompleteCheckout(ctx context.Context, opts CompleteCheckoutOptions) (*Check
...
+		if templateHash == "" {
+			debug.Debug("[app] Template has no hash in ign.json, calculating from content")
+			templateHash = calculateTemplateHash(prep.Template)
```
[the remark this answers is not in the document]
Comment (tacogips) [reply internal/app/checkout.go]: Fixed in commit 75d1185 (PR #17)

Changes made:
- checkout.go now always calculates hash from template content (line 302)
- Added explicit validation that hash is not empty (lines 306-309)
- Comment at line 299-300 explains this ensures consistency with config_init.go

The concern about stale/invalid hashes is resolved since we no longer use prep.IgnJson.Hash.

Resolving this thread.

File: internal/app/update.go [hunk 16]
```diff
@@ -0,0 +1,443 @@
+package app
```
[the remark this answers is not in the document]
Comment (tacogips) [reply internal/app/update.go]: Fixed in commit 4625b0d (merged from PR #19)

Changes made:
- Created internal/app/update_test.go with 983 lines and 31 test cases
- Covers PrepareUpdate, CompleteUpdate, FindVariableChanges, GetNewVariableDefinitions, FilterVariablesForPrompt, ApplyDefaults, FormatVariableChanges
- All tests pass successfully

Resolving this thread.

File: internal/cli/update.go [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+)
+
+func init() {
+	// Flags for update
+	updateCmd.Flags().BoolVarP(&updateForce, "force", "f", false, "Overwrite existing files")
+	updateCmd.Flags().BoolVarP(&updateDryRun, "dry-run", "d", false, "Show what would be generated without writing files")
+	updateCmd.Flags().BoolVarP(&updateVerbose, "verbose", "v", false, "Show detailed processing information")
+}
+
+func runUpdate(cmd *cobra.Command, args []string) error {
```
[the remark this answers is not in the document]
Comment (tacogips) [reply internal/cli/update.go]: Fixed in commit 4625b0d (merged from PR #19)

Changes made:
- Created internal/cli/update_test.go with TestTruncateHash function
- 9 test cases covering: empty string, short hash, boundary cases, normal git hash, long hash, single character
- All tests pass successfully

Resolving this thread.

File: internal/cli/update.go [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+	// Output path defaults to current directory
+	outputPath := "."
+	if len(args) > 0 {
+		outputPath = args[0]
+	}
+
+	// Get GitHub token from environment
+	githubToken := getGitHubToken("")
```
[the remark this answers is not in the document]
Comment (tacogips) [reply internal/cli/update.go]: This issue has been addressed in the current code.

The Long help text at line 33 already includes:
"For private GitHub repositories, set GITHUB_TOKEN environment variable"

This is consistent with the checkout.go approach as noted in the review.

Resolving this thread.
